### PR TITLE
Grid Renderer - Improved appearance of grids

### LINKED ISSUE #3674: [Feature]: Improve visibility of coordinates text over map areas similar in color
### Feature Request

It would make discerning coordinates much easier by eye.

**Context:**
I DM games online screen-sharing a player-mode client MapTool instance on a separate screen in fullscreen mode with grid and coordinates on.
This way, players either don't have to learn to use another app to play, or can play when they do not have access to a computer/laptop (e.g. on a phone). When they want their character to move, they tell me the coordinates and I move the tokens for them.
**Over map areas or backgrounds similar in color (e.g. using the built-in "Sand" texture) to the coordinates text it is hard to see the coordinates, especially when screen-sharing a high resolution window to a low resolution device (e.g. phone).**
Under these circumstances it happens a lot that players have trouble discerning the coordinates.

Resolution probably connected to #3670, #3671, #3672, #3673.

### The Solution you'd like

Ability to change color of coordinates text, and/or ability to change color and/or thickness of edge around characters in coordinates text.

### Alternatives that you've considered.

It occurred to me to make the coordinates text color adaptive to the map area underneath it, but ...

1. ... using the inverse of the colors under the coordinates text characters would probably make it ugly and probably less readable than currently if the map area under the characters is finely detailed and/or colorful.
2. ... using some smart algorithm to choose a color for the characters that would be reasonably visible over their current background (and not ugly) would be probably very hard to come up with (if at all possible), and also unreasonably computationally intensive (especially, considering that it would probably have to be recomputed every time the map is moved or zoomed).

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by bubblobill**
### Identify the Bug or Feature request
progresses #3840, #167
resolves #3674

### Description of the Change
Finally realises the intent of my change for GridLineWidth to GridLineWeight.

Separates the rendering of the grid from the construction of the grid shape(s) to draw.
Square, Iso, and Hex Grid create the shape to draw and pass it back to GridRenderer for actual painting.
Square Grid - changed coordinates to have drop shadow on all sides to improve legibility.

GridRenderer 
- scales drawing stroke according to zone scale.
- bases stroke on size of grid (2% of size) for consistent appearance between grid sizes.
- paints grids with a translucent contrast shadow above 50% zoom.

Did not touch the LibGDX stuff other than a refactor.

### Possible Drawbacks
Should be none.

### Documentation Notes
Improved appearance of grids

### Release Notes

- - -
This change is [Reviewable](https://reviewable.io/reviews/RPTools/maptool/5581)

### CODE REVIEW

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java [hunk 1]
```diff
@@ -14,31 +14,93 @@
...
 
 public class GridRenderer {
-  private final Zone zone;
-  private final ZoneRenderer renderer;
+  private ZoneRenderer renderer;
+  private Zone zone;
+  private static float gridLineWeight;
+  private static float scale;
+  private static float baseWidth = 2f;
+  private static Color[] gridColours;
```
Comment (kwvanderlinde) [inline src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:33]: When the current grid changes (either by modifying a map or loading a different map), the `gridColours` needs to be recalculated.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java [hunk 2]
```diff
@@ -14,31 +14,93 @@
...
+  }
+
   public void renderGrid(Graphics2D g, PlayerView view) {
-    int gridSize = (int) (zone.getGrid().getSize() * renderer.getScale());
-    if (!AppState.isShowGrid() || gridSize < ZoneRendererConstants.MIN_GRID_SIZE) {
+    if (zone == null
+        || !AppState.isShowGrid()
+        || zone.getGrid().getSize() * renderer.getScale() < ZoneRendererConstants.MIN_GRID_SIZE) {
+      if (renderer != null) {
+        setRenderer(renderer);
```
Comment (kwvanderlinde) [inline src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: I don't think this call does anything.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java [hunk 3]
```diff
@@ -14,31 +14,93 @@
...
 
-  public GridRenderer(ZoneRenderer renderer) {
+  GridRenderer(ZoneRenderer renderer) {
+    new MapToolEventBus().getMainEventBus().register(this);
+    setRenderer(renderer);
+  }
+
+  public void setRenderer(ZoneRenderer renderer) {
     this.renderer = renderer;
+    if (renderer.getZone() == null) {
```
Comment (kwvanderlinde) [inline src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Renderers don't have null zones.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java
Comment (kwvanderlinde) [inline src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:1]: What do you think about moving the various `Grid#draw(...)` implementations into `GridRenderer`? It would avoid the need to make half the class static and should simplify some of the state management.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 3 above)
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Sure they do. Take it out and discover a whole new world.

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: No zone during ZR  construction.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:1]: There's a reason this is step one on the journey to prettier grids.

This gets gridRenderer actually painting the various grids.

I intend to construct the grids in a consistent fashion later whilst adding fruit.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 3 above)
Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]:
```java
  public ZoneRenderer(Zone zone) {
    if (zone == null) {
      throw new IllegalArgumentException("Zone cannot be null");
    }
    this.zone = zone;
```
They shouldn't, not even during construction.

Regardless, you can also avoid it by passing the `Zone` to the constructor like many other renderers do and you can dispense with this and other null checks for the zone.

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: This is a minimal-change/initial steps PR. Keeping the previous bullet-proofing and stuff.
I will probably institute your RenderBuddy class down the road. For now, all these checks and stuff, make sure that ZRs without Zones, and Zones without Grids, etc. don't cause screwups. Also, it fails tests if I remove the zone null check.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 2 above)
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Sure it does. If there is no Zone, set the renderer again. Rinse and repeat.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 1 above)
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:33]: Sure, and the GridChanged event actually works. </sarcasm>

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 2 above)
Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Ugh, this is the problem with handling nulls in such a dynamic way.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 3 above)
Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Just checked the tests, they work fine without the null handling.

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Well they were failing on something somewhere until I spammed null handling everywhere.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 1 above)
Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:33]: Creating the `gridColours` array is not at all expensive. You might as well just do it every time in `renderGrid()` and don't worry about caching it.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java [hunk 4]
```diff
@@ -14,31 +14,93 @@
...
+  private static float baseWidth = 2f;
+  private static Color[] gridColours;
 
-  public GridRenderer(ZoneRenderer renderer) {
+  GridRenderer(ZoneRenderer renderer) {
+    new MapToolEventBus().getMainEventBus().register(this);
+    setRenderer(renderer);
+  }
+
+  public void setRenderer(ZoneRenderer renderer) {
```
Comment (kwvanderlinde) [inline src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Method should be `private`.

**Comment by kwvanderlinde** [comment]
As long as the grid colour stuff is made to work, I'm happy with the PR. I'd encourage you to double-check the need for the null checks, but I'm not going to push that too hard.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 1 above)
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java:33]: Oh no. Can't do that.
Step 2. Calculate grid colours based on type of display. Line - dashed - dotted - multiply. Calculate negative colour using perception based calculations. Interpolate colours using easing functions. 
2 a) Make considerable more than 4 painting passes, or 
2 b) Create an image to paint instead, or
2 c) Find exotic things on the internet to try.

File: src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java (hunk 4 above)
Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Nope. Because GridChanged is broken, this will be called in ZR onGridChanged. In fact, I'll do that now and add colour calculation to setRenderer().

Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: In that case, do something a little different: set `renderer` directly in the constructor and leave it at that. Then have a different method (`gridChanged()` or whatever you want to call it) that can udpate the the `GridRenderer` state based on the `renderer` that you already have.

That makes it a lot easier to follow the intent compared to possibly having arbitrary `ZoneRenderer` being injected at any time.

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: changes to grid colour via the grid edit tool should trigger gridchanged event, which triggers ZR to update renderer on grid dependant renderers.

Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Please re-read what I wrote.

You don't need to support setting some arbitrary `ZoneRenderer` since `GridRenderer` _already has the `ZoneRenderer`._ This would suffice instead:
```java
  public void onGridChanged() {
    this.zone = renderer.getZone();
    setGridColours();
  }
```

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Then it will have to wait until not 2:45am.

Comment (kwvanderlinde) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: Oh yeah, get some sleep :sweat_smile:

Comment (bubblobill) [reply src/main/java/net/rptools/maptool/client/ui/zone/renderer/GridRenderer.java]: sorted

**Review by kwvanderlinde — APPROVED** [review]
